protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -383,21 +376,7 @@ impl NetworkBehaviour for Behaviour {
                     .or_default() += 1;
                 self.queued_events.push_back(ToSwarm::Dial { opts });
             }
-            Either::Right(Either::Left(handler::direct::Event::DirectConnectionEstablished)) => {
-                self.queued_events.extend([
-                    ToSwarm::NotifyHandler {
-                        peer_id: event_source,
-                        handler: NotifyHandler::One(relayed_connection_id),
-                        event: Either::Left(
-                            handler::relayed::Command::UpgradeFinishedDontKeepAlive,
-                        ),
-                    },
-                    ToSwarm::GenerateEvent(Event::DirectConnectionUpgradeSucceeded {
-                        remote_peer_id: event_source,
-                    }),
-                ]);
```

Comment:
This snippet will need to happen where we previously instantiated `direct::Handler`.

Reply:
@thomaseizinger Yes right, I removed that as never reached but in fact, now, `Either::Right` can be either unreachable or corresponding to where we previously instantiated `direct::Handler`.

Reply:
Not quite right unfortunately.

We want to emit these events when we established a direction connection. Previously, that was only possible with a dedicated `ConnectionHandler` implementation. Now we can just push these events into `queued_events` inside `handle_established_outbound_connection` where we previously said `return direct::Handler` and now return `dummy::ConnectionHandler`.

If you look at the implementation of `direct::Handler`, it doesn't actually do anything other than emitting an event which is used here again to emit one to the behaviour.

We can just not do this back and forth and simply push these events into `queued_events` directly.

Reply:
@thomaseizinger Ok I think I understand (I hope) now. Before, we used the `on_connection_handler_event` to do a match to know what to push as event into `queued_events`. But now, we don't need that anymore and so we can directly push the event into `queued_events` when we establish the outbound connection while returning `dummy::ConnectionHandler`, it doesn't matter now that the we pushed the event correctly.

Sorry for the misunderstandings, the code is super clear but there are complicated parts and I lack experience on it so a lot of questions must seem stupid to you

Reply:
> the code is super clear

As the author of most of `libp2p-dcutr`, I don't think my code is very clear. Especially the indirection that you just removed in this pull request is very confusing.

> a lot of questions must seem stupid to you

Not at all. I am impressed by how well you understand the code-base @tcoratger especially given the short period of time you have been involved in the project.